### PR TITLE
allow instructions to have labels and be set in circuit.append()

### PR DESCRIPTION
**Pull request by ajavadia**
generalization of gate.label

### CODE REVIEW

**Comment by taalexander** [comment]
We should move in the direction of unifying the circuit and pulse instructions. Right now the pulse instruction has a `name` that it treats as a label. Should we move towards renaming this as label and defining a universal instruction type for both pulse and the circuit?

**Comment by ajavadia** [comment]
Yes I agree 100%. I actually prefer the `name` for this purpose, but I think the way Aer treats name and labels is the current limiting factor. Let's discuss.

File: qiskit/circuit/instruction.py [hunk 1]
```diff
@@ -161,6 +163,26 @@ def params(self, parameters):
                 raise CircuitError("invalid param type {0} in instruction "
                                    "{1}".format(type(single_param), self.name))
 
+    @property
```
Comment (kdk) [inline qiskit/circuit/instruction.py:166]: These can be removed from gate.py

File: qiskit/circuit/quantumcircuit.py [hunk 2]
```diff
@@ -465,6 +466,8 @@ def append(self, instruction, qargs=None, cargs=None):
         if not isinstance(instruction, Instruction) and hasattr(instruction, 'to_instruction'):
             instruction = instruction.to_instruction()
 
+        instruction.label = label
```
Comment (kdk) [inline qiskit/circuit/quantumcircuit.py:469]: Can you add to the docstring that this label will be added to the user's `Instruction` instance?

**Comment by ajavadia** [comment]
There's a problem here that `snapshot` Instruction already has a `label` arg, which collides with the base Instruction having `label`. I think we can't go forward with this unless we change Snapshot to not derive from `Instruction` -- which seems appropriate as it's really not an instruction. Same with barrier.

**Pull request closed without merging**